### ISSUE TITLE
Version negotiation gaps

### ISSUE THREAD

**Issue by MikeBishop** [quicwg/base-drafts#89]
#87 covers improving the current use of ALPN and the Alt-Svc v= parameter for HTTP.  However, I think this is something that QUIC needs to solve generically, so forking off a separate issue for that.

Here are the paths I see that we need to support:

- **Client to server, first time:** This is inherently a 1-RTT handshake *if* the client guesses the version correctly; 2 RTTs if it's wrong.  This incents clients to attempt the most widely-deployed version of QUIC rather than their highest version.  Consequently, there needs to be a way for a conservative client to discover that a server supports a more advanced version.
- **Client to server, subsequent:** The cached information should contain information about the server's supported versions so the client can come back with the same version as last time. The scenarios where one side has changed need consideration, though:
    - **Client upgraded:** Would require client to cache server's full set of supported versions and compare against the client's supported set at connect time, not just caching the selected version.
    - **Server upgraded:** The current mechanism offers no way for a client to discover that the server would have accepted a version it didn't initially offer, whether because the server didn't previously support it or the client is being conservative.

Ideally, I'd like to see these solved for QUIC independent of Alt-Svc -- there will be other users besides HTTP.

I think this could be solved by requiring the server to always send its Version Negotiation list in the crypto handshake, not just if a Version Negotiation packet was actually sent.  Clients can discover when the server supports a newer version than they attempted, and either know what to offer next time or actively open a new connection with the new version in parallel.

**Comment by janaiyengar** [comment]
Yeah, I agree and I think your suggestion of always sending a version list along with the Server Params is a good idea.

**Comment by martinthomson** [comment]
@MikeBishop, I think that we addressed this, do you agree?

**Comment by martinthomson** [comment]
Closing; @MikeBishop open a new issue if we missed something.  I think that we have already covered the main part of this in #122.
